File: android/src/main/java/org/ligi/passandroid/model/PassClassifier.java
package org.ligi.passandroid.model;

import java.util.Collection;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.CopyOnWriteArraySet;

public class PassClassifier {

    public interface OnClassificationChangeListener {
        void OnClassificationChange();
    }

    public Set<OnClassificationChangeListener> onClassificationChangeListeners = new CopyOnWriteArraySet<>();

    public String DEFAULT_TOPIC;

    protected final Map<String, String> topic_by_id;

    private final PassStore passStore;

    public PassClassifier(Map<String, String> topic_by_id, PassStore passStore, String defaultTopic) {
        this.topic_by_id = topic_by_id;
        this.passStore = passStore;
        this.DEFAULT_TOPIC = defaultTopic;

        processDataChange();
    }

    public void processDataChange() {

        final Set<String> keysToRemove = new HashSet<>();

        for (String key : topic_by_id.keySet()) {
            if (passStore.getPassbookForId(key) == null) {
                keysToRemove.add(key);
            }
        }

        for (String key : keysToRemove) {
            topic_by_id.remove(key);
        }

    }


    public void moveToTopic(final Pass pass, final String newTopic) {
        topic_by_id.put(pass.getId(), newTopic);

        processDataChange();

        notifyDataChange();
    }

    public void notifyDataChange() {
        for (OnClassificationChangeListener onClassificationChangeListener : onClassificationChangeListeners) {
            onClassificationChangeListener.OnClassificationChange();
        }
    }


    public Collection<String> getTopics() {
        final Collection<String> res = new HashSet<>();

        res.addAll(topic_by_id.values());

        if (res.isEmpty()) {
            res.add(DEFAULT_TOPIC);
        }

        return res;
    }

    public String getTopic(Pass pass) {
        if (topic_by_id.containsKey(pass.getId())) {
            return topic_by_id.get(pass.getId());
        }

        topic_by_id.put(pass.getId(),DEFAULT_TOPIC);

        return DEFAULT_TOPIC;
    }
}

File: android/src/main/java/org/ligi/passandroid/model/PassClassifier.kt
package org.ligi.passandroid.model

import org.ligi.passandroid.model.pass.Pass

open class PassClassifier(val topicByIdMap: MutableMap<String, String>, private val passStore: PassStore) {

    open fun processDataChange() {
        val topicsToRemove = topicByIdMap.filter { it.value.isEmpty() }.map { it.value }

        topicsToRemove.forEach {
            topicByIdMap.remove(it)
        }
        passStore.notifyChange()
    }

    fun moveToTopic(pass: Pass, newTopic: String) {
        topicByIdMap.put(pass.id, newTopic)

        processDataChange()
    }

    fun getTopics(): Set<String> {
        return topicByIdMap.values.toSet()
    }

    fun getPassListByTopic(topic: String): List<Pass> {
        return topicByIdMap.filter { it.value.equals(topic) }.map { passStore.getPassbookForId(it.key) }.filterNotNull()
    }

    fun getTopic(pass: Pass, default: String): String {
        return getTopic(pass.id, default)
    }

    fun getTopic(id: String, default: String): String {
        val s = topicByIdMap[id]
        if (s != null) {
            return s
        }

        if (!default.isEmpty()) {
            topicByIdMap.put(id, default)
            processDataChange()
        }
        return default
    }

    fun removePass(id: String) {
        topicByIdMap.remove(id)
        processDataChange()
    }

    /*
    useful offsets are -1 and 1 to find the topic right and left from the pass
     */
    fun getTopicWithOffset(pass: Pass, offset: Int): String? {
        val indexOf = getTopics().indexOf(getTopic(pass, ""))
        return getTopics().elementAtOrNull(indexOf + offset)
    }

}
